crates/task-impls/src/quorum_vote.rs

```diff
@@ -70,12 +74,168 @@ struct VoteDependencyHandle<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+            .context(format!(
+                "Proposal's parent missing from storage with commitment: {:?}, proposal view {:?}",
+                justify_qc.date().leaf_commit,
+                proposal.view_number,
+            ))?;
+
+        let (Some(parent_state), _) = consensus_reader.state_and_delta(parent.view_number()) else {
+            bail!("Parent state not found! Consensus internally inconsistent")
+        };
+
+        drop(consensus_reader);
+
+        let (validated_state, state_delta) = parent_state
+            .validate_and_apply_header(
+                &self.instance_state,
+                &parent,
+                &proposal.block_header.clone(),
+                vid_share.data.common.clone(),
+                self.version,
+            )
+            .await
+            .context("Block header doesn't extend the proposal!")?;
+
+        let state = Arc::new(validated_state);
+        let delta = Arc::new(state_delta);
+
+        // TODO Upgrades
+        // Validate the DAC.
+        // let message = if cert.is_valid_cert(Upgrade Certificate) {
+        //     // Validate the block payload commitment for non-genesis DAC.
+        //     if cert.date().payload_commit != proposal.block_header.payload_commitment() {
+        //         warn!(
+        //             "Block payload commitment does not equal da cert payload commitment. View = {}",
+        //             *view
+        //         );
+        //         return false;
+        //     }
+        //     if let Ok(vote) = QuorumVote::<TYPES>::create_signed_vote(
+        //         QuorumData {
+        //             leaf_commit: proposed_leaf.commit(),
+        //         },
+        //         view,
+        //         &public_key,
+        //         &self.private_key
+        //     ) {
+        //         GeneralConsensusMessage::<TYPES>::Vote(vote)
+        //     } else {
+        //         error!("Unable to sign quorum vote!");
+        //         return false;
+        //     }
+        // } else {
+        //     error!(
+        //         "Invalid DAC in proposal! Skipping proposal. {:?} cur view is: {:?}",
+        //         cert, cur_view
+        //     );
+        //     return false;
+        // };
+
+        // Now that we've rounded everyone up, we need to update the shared state and broadcast our events.
+        // We will defer broadcast until all states are updated to avoid holding onto the lock during a network call.
+        let mut consensus_writer = self.consensus.write().await;
+
+        // Note: There's a potential inconsistency here between `self.view_number` and `leaf.data.view_number()`.
+        let view = View {
+            view_inner: ViewInner::Leaf {
+                leaf: proposed_leaf.commit(),
+                state: Arc::clone(&state),
+                delta: Some(Arc::clone(&delta)),
+            },
+        };
+        consensus_writer.update_validated_state_map(self.view_number, view.clone());
```

Comment:
As stated in the comment there is a potential inconsistency.  We should use the view number from proposed leaf instead of `self.view_number`

Reply:
That makes sense to me, it doesn't do that in the existing logic, so I wanted to check, sorry I meant to highlight that in the PR. Changing!